### ISSUE TITLE
[Performance]: vLLM's throughput lags behind llama.cpp for single prompt

### ISSUE THREAD

**Issue by kathirvel-balakrishnan** [vllm-project/vllm#34579]
### Proposal to improve performance

I am using AMD Strix Halo APU. The performance for GPT-OSS-120B model with
1. llama.cpp - 50 tok/sec (stays same with increase in number of prompts)
2. vLLM - 7.5 tok/sec (increases with increase in number of prompts)

What in the architecture can be changed to fix this? The ideal scenario would be to use less than 20 concurrent prompts

### Your current environment (if you think it is necessary)

```text
The output of `python collect_env.py`
```

### Before submitting a new issue...

- [x] Make sure you already searched for relevant issues, and asked the chatbot living at the bottom right corner of the [documentation page](https://docs.vllm.ai/en/latest/), which can answer lots of frequently asked questions.

**Comment by npodbielski** [comment]
I tested it recently and with Qwen 3.6 I have in example:
- llama.cpp: 30t/s
- vllm: 4t/s

Very strange in theory both use ROCm but it seems like vLLM is barely working.

**Comment by hogeheer499-commits** [comment]
Maintainer disclosure: I maintain a separate Strix Halo local LLM guide with raw benchmark evidence. This issue matches what I have seen directionally: `llama.cpp` can be much faster than vLLM on single-prompt Strix Halo paths unless the vLLM workload/concurrency path is very carefully pinned down.

A useful way to split the comparison is:

- single prompt / generation-heavy GGUF: compare against direct `llama-bench` and `llama-server` Vulkan/RADV
- prompt-heavy ingestion: test ROCm/HIP separately
- server workload: report 1/2/4/8/16 parallel requests, TTFT, errors, and aggregate t/s
- vLLM: report whether V1, graph capture, `--enforce-eager`, Triton/CK path, model quant, and block settings are active

In my guide, the measured Strix Halo server split was: Vulkan/RADV won the Qwen3.6 path at 1-4 parallel requests, while Lemonade `llamacpp-rocm` won aggregate throughput at 8-16. Direct Qwen3-Coder rows reached 96-98.5 t/s on Vulkan/RADV, and gpt-oss-120b MXFP4 loaded locally at 55.57 t/s direct `llama.cpp`/Vulkan/RADV.

Raw evidence index: https://github.com/hogeheer499-commits/strix-halo-guide/blob/main/data/headline_claims.csv

I would not treat this as a generic "vLLM is slow" conclusion yet. It needs an apples-to-apples matrix: same hardware, same model file/quant, same prompt/generation length, same concurrency, and explicit vLLM engine/graph/eager settings.

**Comment by npodbielski** [comment]
So where are the arguments for vLLM tests? I tested it myself on Framework Desktop and it was slow to the point of barely being usable. With llama.cpp with MPT you have gain around 50% (60t/s vs 90t/s). Whereas in vLLM you have maybe 27 vs 32 t/s in small contexts. With long contexts it is like 4 vs 6t/s. 

If you are using an agent on Strix Halo compacting coversation can take 40minutes with vLLM.

**Comment by feiyehua** [comment]
Hello @DarkLight1337 @robertgshaw2-redhat @WoosukKwon I think i reproduced the problem. 
On gfx11 and gfx12 it requires 
```
        return (
            _ON_GFX1X
            and (sliding_window == 0 or sliding_window == (-1, -1))
            and (qtype == torch.half or qtype == torch.bfloat16)
            and head_size == 128
            and block_size == 16
            and (gqa_ratio >= 3 and gqa_ratio <= 16)
            and max_seq_len <= 128 * 1024
            and alibi_slopes is None
            and kv_cache_dtype == "auto"
            and sinks is None
        )
```
to enable rocm custom full attention kernel it requires head_dim == 128. (currently aiter is unsupported for these two device families) For gpt-oss 120b, the head_dim is 64, so it fall back to Triton paged attention implementation(vllm/v1/attention/ops/chunked_prefill_paged_decode.py:752).
However, the original `kernel_paged_attention_2d` triton kernel is created for big batch size and lacks kv split support. Therefore when `batch_size` is small, the total occupancy is low and the performance of attention kernel is poor.
I tested on gfx1201 (AMD AI Pro R9700) with Qwen 3.5-9B (head_dim == 256). It also fail to use the custom rocm kernel and fall back to triton kernel. 
Based on my test result, it needs about 100ms to decode 1 token (when batch size = 1, context length = 32K), which is much lower than the theoretical performance (640G/s memory bandwidth, should have about 30 tps).

<img width="1123" height="490" alt="Image" src="https://github.com/user-attachments/assets/eb8baf02-0c49-4977-8470-dffbdc2d496b" />



For each full attention layer, one single attention kernel takes about 6ms, and this takes about 55% for one full forward pass.
As a rough reference, on Nvidia L40 GPU with FlashAttention 2 backend, with split kv support it only takes about 200us. And attention takes about 8% in one single forward pass.

<img width="1692" height="643" alt="Image" src="https://github.com/user-attachments/assets/25347221-a104-41cd-8ade-120f85ff90bd" />

I also tested the performance of sglang in the same scenario. They seem to have a Triton kernel with correct split kv support, and the performance of the triton kernel seems decent. However sglang somehow used GEMM kernels for small batch size and the overall performance seems also far from the ideal one. 

<img width="1544" height="604" alt="Image" src="https://github.com/user-attachments/assets/a39bf9a8-f7c8-4304-b7a4-f1b7eeea95b1" />

So to fix this we can add a split kv kernel in the Triton fallback. 
I'm willing to create a pr for this🤗

Testing env:
vllm commit 2c9c07c85e56c799afffd5a671a8a0bace377a39,
ROCm 7.2.3 / CUDA 12.9, cuda version using prebuild wheel and ROCm compiled from source

Testing script:
```
from __future__ import annotations

import os
import time
from pathlib import Path

import numpy as np
import torch

REPO_ROOT = Path(__file__).resolve().parents[1]
MODEL = "./Qwen3.5-9B"
INPUT_LEN = 32768
OUTPUT_LEN = 257
WARMUP_INPUT_LEN = 32768
WARMUP_OUTPUT_LEN = 8
MAX_MODEL_LEN = 33032
MAX_NUM_BATCHED_TOKENS = 32768
MAX_NUM_SEQS = 1
GPU_MEMORY_UTILIZATION = 0.90
SEED = 0


def _make_prompt(rng: np.random.Generator, length: int) -> dict[str, list[int]]:
    return {"prompt_token_ids": rng.integers(10000, size=length).tolist()}


def main() -> None:
    out_dir = REPO_ROOT / (
        "benchmark_results/"
        f"gdn_vllm_profile_32k_decode256_minimal_{time.strftime('%Y%m%d_%H%M%S')}"
    )
    trace_dir = out_dir / "torch_profiler"
    trace_dir.mkdir(parents=True, exist_ok=True)

    from vllm import LLM, SamplingParams

    profiler_config = {
        "profiler": "torch",
        "torch_profiler_dir": str(trace_dir),
        "torch_profiler_record_shapes": True,
        "torch_profiler_with_stack": False,
        "torch_profiler_with_memory": False,
        "torch_profiler_dump_cuda_time_total": True,
        "torch_profiler_use_gzip": True,
    }

    llm = LLM(
        model=MODEL,
        tokenizer=MODEL,
        trust_remote_code=True,
        dtype="auto",
        tensor_parallel_size=1,
        gpu_memory_utilization=GPU_MEMORY_UTILIZATION,
        max_model_len=MAX_MODEL_LEN,
        max_num_batched_tokens=MAX_NUM_BATCHED_TOKENS,
        max_num_seqs=MAX_NUM_SEQS,
        enable_prefix_caching=False,
        enforce_eager=False,
        async_scheduling=True,
        profiler_config=profiler_config,
        additional_config={"gdn_prefill_backend": "auto"},
    )

    rng = np.random.default_rng(SEED)
    warmup_prompt = _make_prompt(rng, WARMUP_INPUT_LEN)
    measured_prompt = _make_prompt(rng, INPUT_LEN)

    warmup_params = SamplingParams(
        temperature=1.0,
        top_p=1.0,
        ignore_eos=True,
        max_tokens=WARMUP_OUTPUT_LEN,
        detokenize=False,
    )
    measured_params = SamplingParams(
        temperature=1.0,
        top_p=1.0,
        ignore_eos=True,
        max_tokens=OUTPUT_LEN,
        detokenize=False,
    )

    print("Running unprofiled 32K warmup request...")
    llm.generate([warmup_prompt], sampling_params=warmup_params, use_tqdm=False)
    torch.cuda.synchronize()

    print("Running profiled 32K + 257 request...")
    start = time.perf_counter()
    llm.start_profile()
    outputs = llm.generate(
        [measured_prompt],
        sampling_params=measured_params,
        use_tqdm=False,
    )
    llm.stop_profile()
    torch.cuda.synchronize()
    elapsed_s = time.perf_counter() - start
    generated_tokens = len(outputs[0].outputs[0].token_ids)

    print(f"Generated tokens: {generated_tokens}")
    print(f"Profiled request wall time: {elapsed_s:.3f} s")
    print(f"Torch profiler output dir: {trace_dir}")


if __name__ == "__main__":
    main()


```

**Comment by npodbielski** [comment]
@feiyehua is this possible to set batch size to bigger ones via `vllm serve` to use more performant kernel?

**Comment by feiyehua** [comment]
> [@feiyehua](https://github.com/feiyehua) is this possible to set batch size to bigger ones via `vllm serve` to use more performant kernel?

It seems, that, in my case, it will unconditionally fall back to triton attention. However in the big batch size (say, bigger 8) than  cases it should have an acceptable performance.

**Comment by npodbielski** [comment]
Yeah that https://github.com/vllm-project/vllm/issues/45403 
I honestly do not know why this engine is have such good opinions in the internet. Must be just some velocity vllm gained by coming up with paged attention. 
It is slow to start, slow during prompt processing and slow during interference with local models. Of course I do not have a rig with 8 NVIDIA Blackwell GPUs and then it may be where it shines... but for someone playing with local models for fun it is hardly worth it. I almost gave up with local llms till I tried llama.cpp again. On R9700 I have on average 100t/s with Qwen 3.6 MOE. vLLM had barelly 30t/s.

I tried another model and it was unable to start in 10min. Just hanged. 

Then on another try it started in like 5min. And I tried to restart conversation of 80k tokens just to check if this will work. 

It hanged. An timed out after 10mins. I tried to retry. Got bored and restarted it. 

It started after 2min. I retried the conversation and it hanged. 

Best thing is that used scripts that worked like two versions ago. At that point I have no idea why I even bother with it.

**Comment by Lafunamor** [comment]
@feiyehua pointed at the RDNA branch of `use_rocm_custom_paged_attention` above. I have a Strix Halo box (gfx1151 / Radeon 8060S, ROCm 7.14), so I measured that gate rather than reasoning about it. Two findings, one of which contradicts the advice given earlier in this thread.

**1. The `gqa_ratio >= 3` condition is not justified by performance.**

The navi launcher in `csrc/rocm/attention.cu` instantiates `GQA_RATIO` cases 1 and 2 (and the GFX11 kernel has an explicit `if (GQA_RATIO == 1)` branch), so those ratios are dispatchable — the gate is a Python-side policy, not a kernel limit.

Driving `chunked_prefill_paged_decode` both ways at `head_size=128`, `block_size=16`, bf16, comparing the blocked `gqa_ratio=1` against the allowed `gqa_ratio=4` as a control. Speedup is triton/custom, so >1 means the custom kernel wins:

```
  batch   gqa=1 (blocked)   gqa=4 (allowed)
      1         4.61x            4.68x
      2         4.15x            4.09x
      4         3.46x            3.38x
      8         1.40x            1.39x
     16         1.04x            1.03x
     24         0.97x            0.96x
     32         0.96x            0.95x
```

The two columns track each other within ~2% everywhere, and outputs agree numerically (max abs diff ~1e-3 in bf16, same as the allowed ratios). There is no cliff at gqa<3. For reference the gate was introduced in #17004, which set gfx9 to `>= 1` and navi to `>= 3` in the same commit with no comment and no discussion of the choice.

**2. "Use a bigger batch to get the faster kernel" is backwards for this kernel.**

Earlier in the thread the suggestion was that larger batches would reach the more performant path. What the sweep shows is the opposite: the custom kernel's advantage is *largest at batch 1* and disappears around batch 16, after which it is slightly slower than Triton — for **every** gqa_ratio, including the ones currently allowed.

To be clear about what that does and does not mean: aggregate throughput still improves with batching, because you amortise across more sequences. But the *custom attention kernel* is not what makes large batches fast, and the gate's `gqa_ratio` condition does not select for the case where it helps. What predicts the win is batch size, which the gate does not consider at all — `use_rocm_custom_paged_attention` receives no batch argument.

So for the single-prompt case this issue is about, the gate is actively working against you: batch 1 is exactly where the custom kernel is 4.6x faster, and exactly where models with `gqa_ratio` 1 or 2 are refused it.

**Caveats:** gfx1151 only (I cannot test gfx1100 or gfx12xx), decode-only microbenchmark rather than end-to-end tok/s, `head_size=128` / `block_size=16` so the other gate conditions hold, single run per point. Happy to share the script or run more shapes.

I have not opened a PR for this — widening the `gqa_ratio` bound to match gfx9 would be the obvious minimal change, but given finding 2 a batch-aware predicate may be the better fix, and that needs numbers from more than one RDNA part before anyone designs it.

**Comment by npodbielski** [comment]
So do you know what parameters and/or what code changes needs to be done to actually have comparable performance?
